### ISSUE TITLE
Before posting in the issue tracker

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#445]
Thank you for engaging with ty!

Before opening an issue on this repository, note that ty is currently in preview and not ready for production use.

That said, we are very much interested in getting feedback. Before opening a new ticket, please search the issue tracker for existing issues. In particular, please take note of the following known ty issues that users are likely to run into on larger projects:

- **[Missing support for type narrowing on attribute and subscript expressions](https://github.com/astral-sh/ty/issues/164)**
  We do not support narrowing on attribute (`obj.attr`) and subscript (`obj[index]`) expressions, yet.
  ```py
  class C:
      optional_attr: int | None
  
  def f(c: C):
      if c.optional_attr is not None:
          c.optional_attr  # still `int | None` at the moment
  ```

  Common symptoms:
    * False `possibly-unbound-attribute` diagnostics ("**Attribute** `a` on type `C | None` is **possibly unbound**")
    * Various other false diagnostics like `invalid-attribute-type`, with an unexpectedly broad type (**"Expected `str`, found `str | None`"**)

- **[Fatal errors for self-self-referential (recursive) generic classes, type aliases, or protocols](https://github.com/astral-sh/ty/issues/256)**
  ty currently crashes for various self-referential constructs like generic classes, type aliases or protocols. For example:
  ```py
  type C = dict[str, "C"]
  ```

  Common symptoms:
    * ty crashes with *"A fatal error occurred while checking some files"*. The panic message (search for "Panicked at") includes the message **"too many cycle iterations"**
    * Same as above, but the message includes **"dependency graph cycle"** instead.

- **[False positives on constructor calls of generics with explicitly annotated `self` parameters](https://github.com/astral-sh/ty/issues/258)**
  
  ty might issue false positives for constructor calls on generic classes whose `self` argument is explicitly specialized (`self: C[int]`). For example:

  ```py
  import tempfile
  
  # No overload of bound method `__init__` matches arguments (lint:no-matching-overload)
  with tempfile.TemporaryDirectory() as tmp:
      ...
  ```

  Common symptoms:
    *  "`no-matching-overload`: **No overload of bound method `__init__` matches arguments**"
  
- **[No implicit import of submodules](https://github.com/astral-sh/ty/issues/133)**

  If you import `some_module`, ty [will not let you access attributes on submodules](https://play.ty.dev/34083481-1d0a-45b0-80db-04039933cb31) (`some_module.some_submodule.SubmoduleClass`), unless you explicitly import `module.submodule` as well.

  Common symptoms:
    * `unresolved-attribute` diagnostics on modules (*Type `<module 'some_module'>` **has no attribute** `some_submodule`*)
